app/controllers/application_controller.rb

```diff
@@ -1,12 +1,80 @@
 class ApplicationController < ActionController::Base
+  add_flash_types :error, :success, :notice
   # Adds a few additional behaviors into the application controller
   include Blacklight::Controller
   include Fishrappr::Context
+  # Behavior for devise.  Use remote user field in http header for auth.
+  include Devise::Behaviors::HttpHeaderAuthenticatableBehavior
+  
   layout 'blacklight'
 
   # Prevent CSRF attacks by raising an exception.
   # For APIs, you may want to use :null_session instead.
   protect_from_forgery with: :exception
+  
+  before_action :clear_session_user
+
+  # saves the location before loading each page so we can return to the
+  # right page. If we're on a devise page, we don't want to store that as the
+  # place to return to (for example, we don't want to return to the sign in page
+  # after signing in), which is what the :unless prevents
+  before_action :store_current_location, :unless => :devise_controller? || :sessions_controller?
+  
+
   before_action :setup_publication
+
+  # From PSU's ScholarSphere
+  # Clears any user session and authorization information by:
+  #   * ensuring the user will be logged out if REMOTE_USER is not set
+  # 14June17 GML Modified to restore fishrappr session essentials
+  def clear_session_user
+    return nil_request if request.nil?
+    
+    # search = session[:search].dup if session[:search]
+    # publication = session[:publication].dup if session[:publication]
+    user_return_to = session[:user_return_to].dup if session[:user_return_to]
+    mynotice = flash[:notice]
+    
+    request.env['warden'].logout unless user_logged_in?
+    
+    # session[:search] = search
+    # session[:publication] = publication
```

Comment:
Hmmm, I included: @publication = Publication.where(slug: session[:publication]).first
but maybe that can be deleted based on your earlier comment.

Since I made the last round of changes flash doesn't seem to be clearing. Not sure why.